sphinx/environment/__init__.py

```diff
@@ -819,19 +819,28 @@ def _traverse_toctree(
                 traversed.add(sub_docname)
 
 
-def _check_toc_parents(toctree_includes: dict[str, list[str]]) -> None:
+def _check_toc_parents(toctree_includes: dict[str, list[str]], root_doc: str) -> None:
+    """Checks if document is referenced in multiple toctrees.
+    Based on the current implementation of `global_toctree_for_doc`,
+    it considers only the descendants of root_doc and not the whole graph.
+    """
     toc_parents: dict[str, list[str]] = {}
-    for parent, children in toctree_includes.items():
-        for child in children:
-            toc_parents.setdefault(child, []).append(parent)
 
+    def _find_toc_parents_dfs(node: str) -> None:
+        for child in toctree_includes.get(node, []):
+            toc_parents.setdefault(child, []).append(node)
+            is_child_already_visited = len(toc_parents[child]) > 1
+            if not is_child_already_visited:
+                _find_toc_parents_dfs(child)
+
+    _find_toc_parents_dfs(root_doc)
     for doc, parents in sorted(toc_parents.items()):
```

Comment:
A personal opinion: the code might be easier to understand if the iteration source of the subsequent `for` loop -- `toc_parents` -- is assigned-to from the result of the `_find_toc_parents_dfs` function.

Explaining why: to me, function calls that have side-effects that affect outer-scoped variables are slightly hard to follow.

I think that another potential benefit could be that it'd be easier to write test coverage for the helper function (although I admit that it's a small one, and that perhaps the enclosing function is a better candidate for testing here).

Reply:
> A personal opinion: the code might be easier to understand if the iteration source of the subsequent `for` loop -- `toc_parents` -- is assigned-to from the result of the `_find_toc_parents_dfs` function.
> 
> Explaining why: to me, function calls that have side-effects that affect outer-scoped variables are slightly hard to follow.
> 
> I think that another potential benefit could be that it'd be easier to write test coverage for the helper function (although I admit that it's a small one, and that perhaps the enclosing function is a better candidate for testing here).

I didn't worry too much about side-effects as it being more simplistic this way.  

Here is the DFS without side-effect:
```py
    def _find_toc_parents_dfs(node: str, toc_parents: dict[str, list[str]] = {}) -> dict[str, list[str]]:
        for child in toctree_includes.get(node, []):
            already_visited = child in toc_parents
            toc_parents.setdefault(child, []).append(node)
            if already_visited:
                continue
            _find_toc_parents_dfs(child, toc_parents)
        return toc_parents
```
Personally I found it slightly more complicated than needed because of `toc_parents` being propogated down the tree as a parameter but also being returned. Note that `return toc_parents` will only used by the external caller of the helper function and not elsewhere.  
Anyways I'm fine with this implementation too if you think so.